### ISSUE TITLE
Service internal error: Can't remove segment

### ISSUE THREAD

**Issue by yesletgo** [qdrant/qdrant#3080]
## Current Behavior

Service internal error: Can't remove segment data at ./storage/collections/vtl_en/15/segments/34a8bd96-3707-4b1b-be77-644a02fb6e11.deleted, error: Directory not empty (os error 39)"
        },
## Steps to Reproduce

**1.create collection**
qdrant version is 1.6.1 and is cluster mode
kernel: 5.4.248-1.el7.elrepo.x86_64
k8s version: 1.27.4

{
    "vectors": {
      "size": 1024,
      "distance": "Dot",
      "on_disk": false
    },
	"on_disk_payload": false,
    "optimizers_config": {
        "memmap_threshold": 20000
    },
    "quantization_config": {
        "scalar": {
            "type": "int8",
            "always_ram": true
        }
    },
     "hnsw_config": {
        "on_disk": false
    }
}
**2. create index**
```json
{
     "field_name": "group_id",
    "field_schema": "keyword"
}
```

**3.upload  batchs**

**4. result**
{
    "result": {
        "status": "red",
        "optimizer_status": {
            "error": "**Service internal error: Can't remove segment data at ./storage/collections/vtl_en/15/segments/34a8bd96-3707-4b1b-be77-644a02fb6e11.deleted, error: Directory not empty (os error 39)**"
        },
        "vectors_count": 510000,
        "indexed_vectors_count": 130961,
        "points_count": 510000,
        "segments_count": 121,
        "config": {
            "params": {
                "vectors": {
                    "size": 1024,
                    "distance": "Dot",
                    "on_disk": false
                },
                "shard_number": 16,
                "replication_factor": 1,
                "write_consistency_factor": 1,
                "on_disk_payload": false
            },
            "hnsw_config": {
                "m": 16,
                "ef_construct": 100,
                "full_scan_threshold": 10000,
                "max_indexing_threads": 0,
                "on_disk": false
            },
            "optimizer_config": {
                "deleted_threshold": 0.2,
                "vacuum_min_vector_number": 1000,
                "default_segment_number": 0,
                "max_segment_size": null,
                "memmap_threshold": 20000,
                "indexing_threshold": 20000,
                "flush_interval_sec": 5,
                "max_optimization_threads": 1
            },
            "wal_config": {
                "wal_capacity_mb": 32,
                "wal_segments_ahead": 0
            },
            "quantization_config": {
                "scalar": {
                    "type": "int8",
                    "always_ram": true
                }
            }
        },
        "payload_schema": {
            "group_id": {
                "data_type": "keyword",
                "points": 510000
            },
            "cluster": {
                "data_type": "integer",
                "points": 510000
            }
        }
    },
    "status": "ok",
    "time": 0.021438074
}

**Comment by generall** [comment]
Hi @yesletgo, could you please describe how do you mount volumes in your setup?

**Comment by yesletgo** [comment]
**1,  create rbac authorization**

![image](https://github.com/qdrant/qdrant/assets/21998052/a24cff31-9868-4766-b479-5cc3bad852e7)
![image](https://github.com/qdrant/qdrant/assets/21998052/5575bc23-6640-4663-b434-679cbdf4df82)

**2,create Storageclass**

![image](https://github.com/qdrant/qdrant/assets/21998052/e51098ee-ca53-4e53-8d83-e401b1236f86)

**3,create nfs-client-provisioner**

![image](https://github.com/qdrant/qdrant/assets/21998052/472889f5-44b4-4627-b159-30e7d6356963)

finally,
![image](https://github.com/qdrant/qdrant/assets/21998052/e991b978-6598-47d6-a48b-16d295f68503)

**Comment by yesletgo** [comment]
I tested on a network-mounted disk,its performance was fast:
![image](https://github.com/qdrant/qdrant/assets/21998052/73e844ad-b04b-452c-91bb-cad123756448)

**Comment by timvisee** [comment]
I believe that 62.8K IOPS and 245MB/s bandwidth is not that fast.

Are you easily able to reproduce this error every time with the above setup you've described?

Also, whats the reason for having so many PV(C)s? Are you using 12 nodes?

**Comment by yesletgo** [comment]
> I believe that 62.8K IOPS and 245MB/s bandwidth is not that fast.
> 
> Are you easily able to reproduce this error every time with the above setup you've described?
> 
> Also, whats the reason for having so many PV(C)s? Are you using 12 nodes?

Yes, i am using 12 nodes, and this error has occurred frequently recently

**Comment by timvisee** [comment]
> I tested on a network-mounted disk,its performance was fast: ![image](https://user-images.githubusercontent.com/21998052/285080508-73e844ad-b04b-452c-91bb-cad123756448.png)

In this case, you've Qdrant running in a Linux environment right?

I'm asking because the terminal emulator looks Windows-y, and I've seen directory deletion errors like this before on Windows?

**Comment by yesletgo** [comment]
> > log error:
> > ![image](https://user-images.githubusercontent.com/21998052/285081163-ef4c160a-ecf6-494e-94cc-1fa4729bbe16.png)
> 
> I did assume that you're running Qdrant in a Linux environment - but, are you running it on Windows by any chance?
> 
> I'm asking because it looks like I see a Windows based terminal emulator, and I've seen directory deletion errors like this before on Windows[1](#user-content-fn-1-52b0b559878f1c33b8b8188db1ad581f).
> 
> Edit: and indeed I did see this before: [rust-lang/rust#29497 (comment)](https://github.com/rust-lang/rust/issues/29497#issuecomment-260755122) 😄
> 
> ## Footnotes
> 1. https://github.com/rust-lang/rust/issues/29497 [↩](#user-content-fnref-1-52b0b559878f1c33b8b8188db1ad581f)

Never run on Windows.

**Comment by generall** [comment]
I suspect that NFS might be the reason of this issue. I don't think we use NFS in our helm chart, and it have different guarantees compared to regular ext4/other linux FS

**Comment by yesletgo** [comment]
> I suspect that NFS might be the reason of this issue. I don't think we use NFS in our helm chart, and it have different guarantees compared to regular ext4/other linux FS

If installing a qdrant cluster using k8s, how can I avoid using shared storage?For example, using local disk.

**Comment by bashofmann** [comment]
@yesletgo The Qdrant helm chart is not directly using any particular storage solution, but it is relying on Kubernetes StorageClasses and PersistentVolumeClaims, so that Kubernetes can provide persistent storage for the Qdrant Pods. How this is done and what storage is used is transparent for Qdrant itself and solely done by your Kubernetes cluster and how you configured it.
By default the helm charts is using the default storage class that is configured in your Kubernetes cluster. But you can change it to use any other StorageClass that you have available here https://github.com/qdrant/qdrant-helm/blob/main/charts/qdrant/values.yaml#L117-L121 (`persistence.storageClassName`).

NFS is definitely not a good choice for Qdrant. It usually has poor performance and you don't get a proper filesystem which Qdrant relies on. This is not just about Qdrant but also the case for most other database solutions, such es ElasticSearch or Postgres.

Where are you hosting your Kubernetes cluster? What Kubernetes distribution are you using? And what storage solutions do you have available?

Some common examples:
* On AWS (EKS or any other distribution) you would use AWS Elastic Block Storage (EBS): https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html
* On Azure (AKS or any other distribution) you would use Azure Locally Redundant Storage (LRS): https://learn.microsoft.com/en-us/azure/aks/concepts-storage#storage-classes
* On Google Cloud (GKE or any other distribution) you would use Persistent Disks: https://cloud.google.com/kubernetes-engine/docs/concepts/storage-overview#pd
* On VMWare vSphere (Tanzu or any other distribution) you would you vSphere volumes: https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/3.0/vmware-vsphere-csp-getting-started/GUID-221BDDDB-D6A8-4ECE-B275-73A9CF19A7D7.html
* If you have a storage appliance, they usually also have Kubernetes CSI drivers to provision block storage. E.g. NetApp Trident https://netapp-trident.readthedocs.io/en/stable-v19.01/kubernetes/trident-csi.html
* If you don't have any storage setup, there are also solutions that give you replicated, distributed persistent block storage using the disks of the Kubernetes worker nodes, such as Longhorn: https://longhorn.io/

The technologies how the block storage volume is mounted on a node, is different depending on the solution and the implementation of their driver. Often it's something like ISCSI.

As you can see, how to do storage in Kubernetes dependents on your environment. Qdrant just has some requirements (proper file system, not too slow), how you full-fill them is something you have to decide.

**Comment by yesletgo** [comment]
Thank you very much for your answer.

**Comment by jonaskyhse** [comment]
@bashofmann, would it not be ok to use EFS instead of EBS on AWS?

**Comment by bashofmann** [comment]
@jonaskyhse No, it will result in exactly this error, because EFS is not a block storage volume with a proper file-system but AWS' managed NFS product.

**Comment by goaaron** [comment]
> @yesletgo The Qdrant helm chart is not directly using any particular storage solution, but it is relying on Kubernetes StorageClasses and PersistentVolumeClaims, so that Kubernetes can provide persistent storage for the Qdrant Pods. How this is done and what storage is used is transparent for Qdrant itself and solely done by your Kubernetes cluster and how you configured it. By default the helm charts is using the default storage class that is configured in your Kubernetes cluster. But you can change it to use any other StorageClass that you have available here https://github.com/qdrant/qdrant-helm/blob/main/charts/qdrant/values.yaml#L117-L121 (`persistence.storageClassName`).
> 
> NFS is definitely not a good choice for Qdrant. It usually has poor performance and you don't get a proper filesystem which Qdrant relies on. This is not just about Qdrant but also the case for most other database solutions, such es ElasticSearch or Postgres.
> 
> Where are you hosting your Kubernetes cluster? What Kubernetes distribution are you using? And what storage solutions do you have available?
> 
> Some common examples:
> 
> * On AWS (EKS or any other distribution) you would use AWS Elastic Block Storage (EBS): https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html
> * On Azure (AKS or any other distribution) you would use Azure Locally Redundant Storage (LRS): https://learn.microsoft.com/en-us/azure/aks/concepts-storage#storage-classes
> * On Google Cloud (GKE or any other distribution) you would use Persistent Disks: https://cloud.google.com/kubernetes-engine/docs/concepts/storage-overview#pd
> * On VMWare vSphere (Tanzu or any other distribution) you would you vSphere volumes: https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/3.0/vmware-vsphere-csp-getting-started/GUID-221BDDDB-D6A8-4ECE-B275-73A9CF19A7D7.html
> * If you have a storage appliance, they usually also have Kubernetes CSI drivers to provision block storage. E.g. NetApp Trident https://netapp-trident.readthedocs.io/en/stable-v19.01/kubernetes/trident-csi.html
> * If you don't have any storage setup, there are also solutions that give you replicated, distributed persistent block storage using the disks of the Kubernetes worker nodes, such as Longhorn: https://longhorn.io/
> 
> The technologies how the block storage volume is mounted on a node, is different depending on the solution and the implementation of their driver. Often it's something like ISCSI.
> 
> As you can see, how to do storage in Kubernetes dependents on your environment. Qdrant just has some requirements (proper file system, not too slow), how you full-fill them is something you have to decide.

How does using `EBS` PVC work with Qdrant in distributed mode? Would EFS not be the choice in this scenario?
